README.md
# [arc4](https://github.com/hex7c0/arc4)

[![NPM version](https://img.shields.io/npm/v/arc4.svg)](https://www.npmjs.com/package/arc4)
[![Linux Status](https://img.shields.io/travis/hex7c0/arc4.svg?label=linux)](https://travis-ci.org/hex7c0/arc4)
[![Windows Status](https://img.shields.io/appveyor/ci/hex7c0/arc4.svg?label=windows)](https://ci.appveyor.com/project/hex7c0/arc4)
[![Dependency Status](https://img.shields.io/david/hex7c0/arc4.svg)](https://david-dm.org/hex7c0/arc4)
[![Coveralls](https://img.shields.io/coveralls/hex7c0/arc4.svg)](https://coveralls.io/r/hex7c0/arc4)

[RC4](https://en.wikipedia.org/wiki/RC4) stream cipher.
You can select from ["[arc4](https://en.wikipedia.org/wiki/RC4)", "[rc4a](https://en.wikipedia.org/wiki/RC4#RC4A)", "[vmpc](https://en.wikipedia.org/wiki/RC4#VMPC)", "[rc4+](https://en.wikipedia.org/wiki/RC4#RC4.2B)"] algorithm

Encode/decode with different [encodings](http://nodejs.org/api/buffer.html#apicontent) _for *String only_, from nodejs doc:
> - 'ascii' - for 7 bit ASCII data only. This encoding method is very fast, and will strip the high bit if set.
> - 'utf8' - Multibyte encoded Unicode characters. Many web pages and other document formats use UTF-8.
> - 'utf16le' - 2 or 4 bytes, little endian encoded Unicode characters. Surrogate pairs (U+10000 to U+10FFFF) are supported.
> - 'ucs2' - Alias of 'utf16le'.
> - 'base64' - Base64 string encoding.
> - 'binary' - A way of encoding raw binary data into strings by using only the first 8 bits of each character. This encoding method is deprecated and should be avoided in favor of Buffer objects where possible. This encoding will be removed in future versions of Node.
> - 'hex' - Encode each byte as two hexadecimal characters.

My original [python code](https://github.com/hex7c0/EncryptoPy/blob/master/modules/rc/rc4.py)

## Installation

Install through NPM

```bash
npm install arc4
```
or
```bash
git clone git://github.com/hex7c0/arc4.git
```

## API

inside nodejs project
```js
var rc4 = require('arc4');

var cipher = rc4('arc4', 'secret_key');
var d = cipher.encodeString('ciao');
var e = cipher.decodeString(d);
```

### Methods

change your key and reload [gKsa](https://en.wikipedia.org/wiki/RC4#Key-scheduling_algorithm_.28KSA.29) (warning)
```js
cipher.change('foo');
cipher.change([30, 31]);
cipher.change(new Buffer('foo'));
```

encode a plaintext string, you can optionally choose input (defaults to 'utf8') and output (defaults to 'hex') [encoding](http://nodejs.org/api/buffer.html#apicontent)
```js
cipher.encodeString(plaintext [, input_encoding [, output_encoding]]);
cipher.encodeString('string', 'utf8', 'base64');
```

encode a plaintext array
```js
cipher.encodeArray([49, 50, 51]);
```

encode a plaintext buffer data
```js
cipher.encodeBuffer(new Buffer('ciao'));
```

select right function according with plaintext data type. Set input and output [encoding](http://nodejs.org/api/buffer.html#apicontent) only if data is a String
```js
cipher.encode(your_data [, input_encoding [, output_encoding]]);
```

decode a ciphertext string, you can optionally choose input (defaults to 'hex') and output (defaults to 'utf8') [encoding](http://nodejs.org/api/buffer.html#apicontent)
```js
cipher.decodeString(ciphertext [, input_encoding [, output_encoding]]);
cipher.decodeString('string', 'utf8', 'base64');
```

decode a ciphertext array
```js
cipher.decodeArray([49,50,51]);
```

decode a ciphertext buffer data
```js
cipher.decodeBuffer(new Buffer('ciao'));
```

select right function according with ciphertext data type. Set input and output [encoding](http://nodejs.org/api/buffer.html#apicontent) only if data is a String
```js
cipher.decode(your_data [, input_encoding [, output_encoding]]);
```

### rc4(algorithm, password [, lodash])

#### algorithm

 - `algorithm` - **String** Choose between ["[arc4](https://en.wikipedia.org/wiki/RC4)", "[rc4a](https://en.wikipedia.org/wiki/RC4#RC4A)", "[vmpc](https://en.wikipedia.org/wiki/RC4#VMPC)", "[rc4+](https://en.wikipedia.org/wiki/RC4#RC4.2B)"] *(default "throw Error")*

#### password

 - `password` - **String | Array | Buffer** Your key *(default "throw Error")*

#### [lodash]

 - `lodash` - **Boolean** Use [lodash](http://lodash.com/) library (check [benchmark](test/benchmark.js) test for right decision) *(default "disabled")*

## Examples

Take a look at my [examples](examples)

### [License GPLv3](license)


History.md
v3.3.4 / 2016-05-21
==================

  * Fix mistyped property of key (@aucelum)
  * Update `lodash`@4.12.0

v3.3.3 / 2016-05-03
==================

  * Tested against `node`@6
  * Update `lodash`@4.11.2

v3.3.2 / 2016-04-12
==================

  * Update `lodash`@4.10.0

v3.3.1 / 2016-03-29
==================

  * Tested against `node`@5.7
  * Update `lodash`@4.6.1

v3.3.0 / 2016-01-13
==================

  * Tested against `node`@5.3
  * Update `lodash`@4.0.0

v3.2.1 / 2015-10-31
==================

  * Tested against `node`@5

v3.2.0 / 2015-08-29
==================

  * files reorganization
  * `iojs@3` test
  * Update `lodash`@3.10.1

v3.1.3 / 2015-07-14
==================

  * Update `lodash`@3.10.0

v3.1.2 / 2015-06-09
==================

  * Fix SPDX license
  * Update `lodash`@3.9.3

v3.1.1 / 2015-04-18
==================

  * Update `lodash`@3.7.0

v3.1.0 / 2015-03-27
==================

  * Remove `.code*` support
  * Fix deprecation error with `.codeString`
  * Update `lodash`@3.6.0
  * `windows` test
  * `iojs` test
  * `coveralls` test

v3.0.8 / 2015-01-30
==================

  * Fix documentation
  * Update `lodash`@3.0.0

v3.0.7 / 2015-01-20
==================

  * Package fix

v3.0.6 / 2014-10-26
==================

  * `jshint`

v3.0.5 / 2014-10-04
==================

  * Remove jsdoc
  * `package.json` min

v3.0.4 / 2014-09-09
==================

  * Performance improvement

v3.0.3 / 2014-09-08
==================

  * Display "deprecated" warning for `codeString`

v3.0.2 / 2014-09-07
==================

  * Code rewrite
  * Add "algorithm" (options)
  * Rename "key" to password (options)
  * Now, every algorithm is called inside class, not with methods (codeArrayRC4A)
  * Add "encode*" and "decode*" methods for every data input
  * Add "input_encoding" and "output_encoding" for "encodeString" and "decodeString"

v2.2.5 / 2014-09-06
==================

  * Testing input_encoding (normal string only)
  * Testing output_encoding (normal string only)

v2.2.3 / 2014-08-23
==================

  * Performance tips

v2.2.2 / 2014-08-11
==================

  * Update README.md

v2.2.1 / 2014-08-06
==================

  * `lodash` library
  * Add "lodash" (options)

v2.1.4 / 2014-08-04
==================

  * Doc update

v2.1.2 / 2014-07-27
==================

  * Test for "file"

v2.1.0 / 2014-07-27
==================

  * Test "node": ">=0.10.0" only
  * Using task runner `grunt`
  * Using test framework `mocha`
  * Testing script will be put inside "test/"
  * ".npmignore" more aggressive
  * `uglify` compiles
  * `RC4p` rename to `RC4P`

v2.0.0 / 2014-07-10
==================

  * Rename "codeByte" to "coreArray"
  * New structure of code

v1.2.0 / 2014-07-08
==================

  * `code()` fix
  * Buffer methods

v1.1.1 / 2014-06-29
==================

  * Remove anonymous function

v1.1.0 / 2014-06-24
==================

  * New method `RC4A`
  * New method `VMPC`
  * New method `RC4+`

v1.0.0 / 2014-06-23
==================

  * Project start


lib/normal.js
"use strict";
/**
 * @file normal main
 * @module arc4
 * @package arc4
 * @subpackage lib
 * @version 2.2.5
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
// load
var box = [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
        22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41,
        42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61,
        62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81,
        82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101,
        102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
        118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133,
        134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149,
        150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165,
        166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181,
        182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197,
        198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213,
        214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229,
        230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245,
        246, 247, 248, 249, 250, 251, 252, 253, 254, 255 ];

/*
 * functions
 */
/**
 * export class
 * 
 * @exports normal
 * @function normal
 * @params {String} key - user key
 * @return {RC4}
 */
module.exports = function normal(key) {

    return new RC4(key);
};

/**
 * generate ksa
 * 
 * @function gKsa
 * @param {Array} key - user key
 * @return {Array}
 */
function gKsa(key) {

    var j = 0;
    var s = box.slice();
    var ke = key;
    var len = ke.length;
    for (var i = 0; i < 256; i++) {
        j = (j + s[i] + ke[i % len]) % 256;
        s[j] = [ s[i], s[i] = s[j] ][0];
    }
    return s;
}

/**
 * body cipher
 * 
 * @function body
 * @param {String|Array|Buffer} inp - input
 * @param {String|Array|Buffer} res - response
 * @param {Array} ksa - ksa box
 * @return {String|Array|Buffer}
 */
function body(inp, res, ksa) {

    var i = 0, j = 0;
    var out = res;
    // if (typeof (res) === 'string') {
    // for (var y = 0, l = inp.length; y < l; y++) {
    // i = (i + 1) % 256;
    // j = (j + ksa[i]) % 256;
    // ksa[j] = [ ksa[i], ksa[i] = ksa[j] ][0];
    // out += String.fromCharCode(inp.charCodeAt(y) ^ ksa[(ksa[i] + ksa[j]) %
    // 256]);
    // }
    // } else {
    for (var y = 0, l = inp.length; y < l; y++) {
        i = (i + 1) % 256;
        j = (j + ksa[i]) % 256;
        ksa[j] = [ ksa[i], ksa[i] = ksa[j] ][0];
        out[y] = inp[y] ^ ksa[(ksa[i] + ksa[j]) % 256];
    }
    // }
    return out;
}

/**
 * body cipher RC4A
 * 
 * @function bodyRC4A
 * @param {String|Array|Buffer} inp - input
 * @param {String|Array|Buffer} res - response
 * @param {Array} ksa - ksa box
 * @return {String|Array|Buffer}
 */
function bodyRC4A(inp, res, ksa) {

    var i = 0, j1 = 0, j2 = 0;
    var s2 = ksa.slice();
    var out = res;
    if (typeof (res) === 'string') {
        for (var y = 0, l = inp.length; y < l; y++) {
            i = (i + 1) % 256;
            j1 = (j1 + ksa[i]) % 256;
            ksa[j1] = [ ksa[i], ksa[i] = ksa[j1] ][0];
            out += String.fromCharCode(inp.charCodeAt(y) ^ s2[(ksa[i] + ksa[j1]) % 256]);
            y++;
            j2 = (j2 + s2[i]) % 256;
            s2[j2] = [ s2[i], s2[i] = s2[j2] ][0];
            out += String.fromCharCode(inp.charCodeAt(y) ^ ksa[(s2[i] + s2[j2]) % 256]);
        }
    } else {
        for (var y = 0, l = inp.length; y < l; y++) {
            i = (i + 1) % 256;
            j1 = (j1 + ksa[i]) % 256;
            ksa[j1] = [ ksa[i], ksa[i] = ksa[j1] ][0];
            out[y] = inp[y] ^ s2[(ksa[i] + ksa[j1]) % 256];
            y++;
            j2 = (j2 + s2[i]) % 256;
            s2[j2] = [ s2[i], s2[i] = s2[j2] ][0];
            out[y] = inp[y] ^ ksa[(s2[i] + s2[j1]) % 256];
        }
    }
    return out;
}

/**
 * body cipher VMPC
 * 
 * @function bodyVMPC
 * @param {String|Array|Buffer} inp - input
 * @param {String|Array|Buffer} res - response
 * @param {Array} ksa - ksa box
 * @return {String|Array|Buffer}
 */
function bodyVMPC(inp, res, ksa) {

    var i = 0, j = 0;
    var a, b;
    var out = res;
    if (typeof (res) === 'string') {
        for (var y = 0, l = inp.length; y < l; y++) {
            a = ksa[i];
            j = ksa[(j + a) % 256];
            b = ksa[j];
            out += String.fromCharCode(inp.charCodeAt(y) ^ ksa[ksa[b] + 1]);
            ksa[j] = [ a, ksa[i] = b ][0];
            i = (i + 1) % 256;
        }
    } else {
        for (var y = 0, l = inp.length; y < l; y++) {
            a = ksa[i];
            j = ksa[(j + a) % 256];
            b = ksa[j];
            out[y] = inp[y] ^ ksa[ksa[b] + 1];
            ksa[j] = [ a, ksa[i] = b ][0];
            i = (i + 1) % 256;
        }
    }
    return out;
}

/**
 * body cipher RC4P
 * 
 * @function bodyRC4P
 * @param {String|Array|Buffer} inp - input
 * @param {String|Array|Buffer} res - response
 * @param {Array} ksa - ksa box
 * @return {String|Array|Buffer}
 */
function bodyRC4P(inp, res, ksa) {

    var i = 0, j = 0;
    var a, b, c;
    var out = res;
    if (typeof (res) === 'string') {
        for (var y = 0, l = inp.length; y < l; y++) {
            i = (i + 1) % 256;
            a = ksa[i];
            j = ksa[(j + a) % 256];
            b = ksa[j];
            ksa[j] = [ a, ksa[i] = b ][0];
            c = (ksa[i << 5 ^ j >> 3] + ksa[j << 5 ^ i >> 3]) % 256;
            out += String.fromCharCode(inp.charCodeAt(y) ^ (ksa[a + b] + ksa[c ^ 0xAA])
                    ^ ksa[j + b]);
        }
    } else {
        for (var y = 0, l = inp.length; y < l; y++) {
            i = (i + 1) % 256;
            a = ksa[i];
            j = ksa[(j + a) % 256];
            b = ksa[j];
            ksa[j] = [ a, ksa[i] = b ][0];
            c = (ksa[i << 5 ^ j >> 3] + ksa[j << 5 ^ i >> 3]) % 256;
            out[y] = inp[y] ^ (ksa[a + b] + ksa[c ^ 0xAA]) ^ ksa[j + b];
        }
    }
    return out;
}

/*
 * class
 */
/**
 * RC4 class
 * 
 * @class RC4
 * @param {String|Array|Buffer} key - user key
 * @return {Object}
 */
function RC4(key) {

    this.key;
    this.change(key);
}
/**
 * change user key
 * 
 * @function change
 * @param {String|Array|Buffer} key - user key
 */
RC4.prototype.change = function(key) {

    this.key = new Array(key.legth);
    if (Array.isArray(key)) {
        this.key = key;
    } else if (typeof (key) === 'string' || Buffer.isBuffer(key)) {
        var key = new Buffer(key);
        for (var i = 0, ii = key.length; i < ii; i++) {
            this.key[i] = key[i];
        }
    } else {
        throw new Error('Invalid data');
    }
    return;
};
/**
 * RC4 string code
 * 
 * @function codeString
 * @param {String} str - data
 * @param {String} input_encoding - input
 * @param {String} output_encoding - output
 * @return {String}
 */
RC4.prototype.codeString = function(str, input_encoding, output_encoding) {

    var out = new Buffer(str, input_encoding || 'binary');
    return body(out, new Buffer(str.length), gKsa(this.key)).toString(
            output_encoding || 'binary');
};
/**
 * RC4 array code
 * 
 * @function codeArray
 * @param {Array} arr - data
 * @return {Array}
 */
RC4.prototype.codeArray = function(arr) {

    return body(arr, new Array(arr.length), gKsa(this.key));
};
/**
 * RC4 buffer code
 * 
 * @function codeBuffer
 * @param {Buffer} buff - data
 * @return {Buffer}
 */
RC4.prototype.codeBuffer = function(buff) {

    return body(buff, new Buffer(buff.length), gKsa(this.key));
};
/**
 * RC4 mixed code. Alias for codeString or codeByte
 * 
 * @function code
 * @param {String|Array|Buffer} boh - data
 * @param {String} input_encoding - input
 * @param {String} output_encoding - output
 * @return {String|Array|Buffer}
 */
RC4.prototype.code = function(boh, input_encoding, output_encoding) {

    if (typeof (boh) === 'string') {
        return this.codeString(boh, input_encoding, output_encoding);
    }
    if (Array.isArray(boh)) {
        return this.codeArray(boh);
    }
    if (Buffer.isBuffer(boh)) {
        return this.codeBuffer(boh);
    }
    throw new Error('Invalid data');
    return;
};
/**
 * RC4A string code
 * 
 * @function codeStringRC4A
 * @param {String} str - data
 * @return {String}
 */
RC4.prototype.codeStringRC4A = function(str) {

    return bodyRC4A(str, '', gKsa(this.key));
};
/**
 * RC4A array code
 * 
 * @function codeArrayRC4A
 * @param {Array} arr - data
 * @return {Array}
 */
RC4.prototype.codeArrayRC4A = function(arr) {

    return bodyRC4A(arr, new Array(arr.length), gKsa(this.key));
};
/**
 * RC4A buffer code
 * 
 * @function codeBufferRC4A
 * @param {Buffer} buff - data
 * @return {Buffer}
 */
RC4.prototype.codeBufferRC4A = function(buff) {

    return bodyRC4A(buff, new Buffer(buff.length), gKsa(this.key));
};
/**
 * RC4A mixed code. Alias for codeString or codeByte
 * 
 * @function codeRC4A
 * @param {String|Array|Buffer} boh - data
 * @return {String|Array|Buffer}
 */
RC4.prototype.codeRC4A = function(boh) {

    if (typeof (boh) === 'string') {
        return this.codeStringRC4A(boh);
    }
    if (Array.isArray(boh)) {
        return this.codeArrayRC4A(boh);
    }
    if (Buffer.isBuffer(boh)) {
        return this.codeBufferRC4A(boh);
    }
    throw new Error('Invalid data');
    return;
};
/**
 * VMPC string code
 * 
 * @function codeStringVMPC
 * @param {String} str - data
 * @return {String}
 */
RC4.prototype.codeStringVMPC = function(str) {

    return bodyVMPC(str, '', gKsa(this.key));
};
/**
 * VMPC array code
 * 
 * @function codeArrayVMPC
 * @param {Array} arr - data
 * @return {Array}
 */
RC4.prototype.codeArrayVMPC = function(arr) {

    return bodyVMPC(arr, new Array(arr.length), gKsa(this.key));
};
/**
 * VMPC buffer code
 * 
 * @function codeBufferVMPC
 * @param {Buffer} buff - data
 * @return {Buffer}
 */
RC4.prototype.codeBufferVMPC = function(buff) {

    return bodyVMPC(buff, new Buffer(buff.length), gKsa(this.key));
};
/**
 * VMPC mixed code. Alias for codeString or codeByte
 * 
 * @function codeVMPC
 * @param {String|Array} boh - data
 * @return {String|Array}
 */
RC4.prototype.codeVMPC = function(boh) {

    if (typeof (boh) === 'string') {
        return this.codeStringVMPC(boh);
    }
    if (Array.isArray(boh)) {
        return this.codeArrayVMPC(boh);
    }
    if (Buffer.isBuffer(boh)) {
        return this.codeBufferVMPC(boh);
    }
    throw new Error('Invalid data');
    return;
};
/**
 * RC4P string code
 * 
 * @function codeStringRC4P
 * @param {String} str - data
 * @return {String}
 */
RC4.prototype.codeStringRC4P = function(str) {

    return bodyRC4P(str, '', gKsa(this.key));
};
/**
 * RC4P array code
 * 
 * @function codeArrayRC4P
 * @param {Array} arr - data
 * @return {Array}
 */
RC4.prototype.codeArrayRC4P = function(arr) {

    return bodyRC4P(arr, new Array(arr.length), gKsa(this.key));
};
/**
 * RC4P buffer code
 * 
 * @function codeBufferRC4P
 * @param {Buffer} buff - data
 * @return {Buffer}
 */
RC4.prototype.codeBufferRC4P = function(buff) {

    return bodyRC4P(buff, new Buffer(buff.length), gKsa(this.key));
};
/**
 * RC4P mixed code. Alias for codeString or codeByte
 * 
 * @function codeRC4P
 * @param {String|Array|Array} boh - data
 * @return {String|Array|Array}
 */
RC4.prototype.codeRC4P = function(boh) {

    if (typeof (boh) === 'string') {
        return this.codeStringRC4P(boh);
    }
    if (Array.isArray(boh)) {
        return this.codeArrayRC4P(boh);
    }
    if (Buffer.isBuffer(boh)) {
        return this.codeBufferRC4P(boh);
    }
    throw new Error('Invalid data');
    return;
};
